libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt

```diff
@@ -75,22 +51,23 @@ class RouteLineComponent(
         }
     }
 
-    private var mapboxNavigation: MapboxNavigation? = null
-
     override fun onAttached(mapboxNavigation: MapboxNavigation) {
         super.onAttached(mapboxNavigation)
-        this.mapboxNavigation = mapboxNavigation
-        contract = contractProvider.get()
         mapPlugins.gestures.addOnMapClickListener(onMapClickListener)
         mapPlugins.location.addOnIndicatorPositionChangedListener(onPositionChangedListener)
 
-        coroutineScope.launch {
-            mapboxNavigation.flowRouteProgress().collect { routeProgress ->
-                ifNonNull(mapboxMap.getStyle()) { style ->
-                    routeLineApi.updateWithRouteProgress(routeProgress) { result ->
-                        routeLineView.renderRouteLineUpdate(style, result).also {
-                            result.error?.let {
-                                Log.e(TAG, it.errorMessage, it.throwable)
+        when (contract) {
+            is MapboxNavigationObserver -> {
```

Comment:
I thought the whole idea behind `Contracts` was to make them simple and to serve as a "glue" code that allows the use of the components in any context.

I don't think components should be responsible for checking if contracts are `MapboxNavigationObservers` and attaching them to `MapboxNavigation` instance. That responsibility should fall on whoever is providing the contract.

Reply:
Yeah I'm not in love with this either. Playing with trade-offs.

The current contract has no way to observe the route, it only allows you to set it. Using the RouteLineComponent becomes more complicated downstream if it cannot observe the routes in the contract

Reply:
Isn't this where the other ticket I mentioned comes into play. When setting the routes using the contract, the customer might be interested in knowing about this change. https://github.com/mapbox/navigation-sdks/issues/1436 and https://github.com/mapbox/navigation-sdks/issues/1645. 
I think this is what you are looking for.

Reply:
I think the issue is different.. But there are a few approaches